### PR TITLE
Update to 0.16 (1.6.1dev, API 2.1.0)

### PR DESCRIPTION
**Pull request by dktapps**
Will document changes here. THIS IS NOT READY FOR MERGE.
### API changes
- Lots of entity metadata flags disappeared and/or changed. There are also a large number of new ones. If you work with metadata in plugins, you may need to update them.
- New Entity API methods (see diff for details)
- PlayerToggleFlightEvent, Player:: isFlying() and Player::setFlying() have been added.
- Giant stack of new entity generic status flags (see diff)
### Known bugs
- Self-duplication in player list (argh!)
- Entity metadata flags higher than 31 will not work on 32-bit systems
- ~~Players can cheat in the GUI with things like Always Day when only one player is online.~~ (Client bug! https://bugs.mojang.com/browse/MCPE-17617)
### TODO
- Varint64 support for 32-bit (or cut support for 32-bit, vote at https://twitter.com/dktapps/status/789199456801402881)
- Resource and behaviour pack support
- New commands API for the new-style 0.16 commands
- Proper support for separate UniqueID and RuntimeID on entities
- Further implementation of vanilla-style permissions

### CODE REVIEW

**Comment by TheAppleGamer** [comment]
Works okay on my end

**Comment by TDMidkwhatisxd** [comment]
I tested it! 
Map load: u need to enter in unloaded part of map, for it to load!
PvP: If a player enter in your zone , and if your finger is on player , you attack him , even if u dont, (hard to explain :D)

Craft work! :3

**Review by PrimusLV — APPROVED** [review]
It's still buggy but compatible with 0.16

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -309,10 +314,19 @@ public function setAllowFlight($value){
...
+	public function getAllowFlight() : bool{
 		return $this->allowFlight;
 	}
 
+	public function setFlying(bool $value){
+		$this->isFlying = $value;
+		$this->sendSettings();
+	}
+
+	public function getIsFlying() : bool{
```
Comment (SOF3) [inline src/pocketmine/Player.php]: Why not isFlying()?

File: src/pocketmine/Player.php [hunk 2]
```diff
@@ -489,6 +504,23 @@ public function getEffectivePermissions(){
...
+		$pk = new AvailableCommandsPacket();
+		$data = new \stdClass();
+		foreach($this->server->getCommandMap()->getCommands() as $command){
+			if(!$command->testPermissionSilent($this)){
+				continue;
+			}
+			$data->{$command->getName()}->versions[0] = $command->generateCustomCommandData($this);
+		}
+
+		if(count($data) > 0){
```
Comment (SOF3) [inline src/pocketmine/Player.php]: Does `count(stdClass)` really work? Are you sure?

**Review by SOF3 — DISMISSED** [review]

**Comment by SOF3** [comment]
There are some other issues handled in 8c87e73468a126788ec4b9b237db6df512604d91 (mixed with some changes from master branch because it is a merge commit).

**Comment by Dog2puppy** [comment]
How do the / commands work exactly?

**Comment by remotevase** [comment]
@Dog2puppy you may want to take a look at LBSG for how the commands look in 0.16 servers.

**Comment by PrimusLV** [comment]
Don't drop 32 bit :)

**Comment by Dog2puppy** [comment]
@PrimusLV We should do what @remotevase did with IM. The official one is 64-bit, but then we could start a 32-bit!

**Comment by dktapps** [comment]
Next person who comes round here advertising other softwares is going to receive a ban. THIS IS NOT A FORUM.

**Comment by HimbeersaftLP** [comment]
How was that advertising? He was just giving a hint...

**Comment by TDMidkwhatisxd** [comment]
Ok , sorry ! :D.
@dktapps , i experience slow fps  , in some zones , even with 4 chunks radius setting!
If i  look on that "gliched or buggy" location , my screen move in slow motion!
I dont know if is usefull , but im reporting!

My phone : GoClever Quantum 1010N , 1gb ram , 4 cpu cores! (I had 480mb free ram wen i joined game!)

**Comment by Misteboss** [comment]
Raklib block the player IP on join
6m[RakLibServer thread/NOTICE]: Blocked 88.182.247.198 for 5 seconds[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Could not tick level "ZenBuildHub": Call to a member function getValue() on null[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Error: "Call to a member function getValue() on null" (EXCEPTION) in "/src/pocketmine/entity/Human" at line 113[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Could not tick level "ZenBuildHub": Call to a member function getValue() on null[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Error: "Call to a member function getValue() on null" (EXCEPTION) in "/src/pocketmine/entity/Human" at line 113[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Could not tick level "ZenBuildHub": Call to a member function getValue() on null[m
[36m[03:22:26] [m[31m[Server thread/CRITICAL]: Error: "Call to a member function getValue() on null" (EXCEPTION) in "/src/pocketmine/entity/Human" at line 113[m
Voila

**Comment by dktapps** [comment]
@Misteboss did you READ my comment on your issue already?

File: src/pocketmine/Player.php [hunk 3]
```diff
@@ -1170,7 +1170,7 @@ public function sendSettings(){
 		$pk->allowFlight = $this->allowFlight;
 		$pk->noClip = $this->isSpectator();
 		$pk->isFlying = $this->flying;
-		$pk->userPermission = 2;
+		$pk->userPermission = ($this->isOp() ? AdventureSettingsPacket::PERMISSION_OPERATOR : AdventureSettingsPacket::PERMISSION_NORMAL);
```
Comment (SOF3) [inline src/pocketmine/Player.php]: You don't really need to parenthesize it.

**Comment by Creeperface01** [comment]
Where is @Intyre ?

**Comment by ZENMCPE** [comment]
How can i download it

**Review by sergeysova — APPROVED** [review]

**Pull request merged**